pkg/controller/azvolumeattachment.go

```diff
@@ -0,0 +1,546 @@
+		}
+	}
+	return reconcile.Result{Requeue: true}, nil
+}
+
+func (r *reconcileAzVolumeAttachment) HandleReplicaUpdate(ctx context.Context, request reconcile.Request) (reconcile.Result, error) {
+	storageClassName := request.Name
+	replicaCount, err := r.GetReplicaCount(ctx, storageClassName)
+	if err != nil {
+		return reconcile.Result{}, err
+	}
+
+	for volume := range r.scMap[storageClassName] {
+		if int(replicaCount) != len(r.scMap[storageClassName]) {
+			err := r.ManageReplicaForVolume(ctx, volume)
+			if err != nil {
+				return reconcile.Result{Requeue: true}, err
+			}
+		}
+	}
+	return reconcile.Result{}, nil
+}
+
+func (r *reconcileAzVolumeAttachment) GetReplicaCount(ctx context.Context, name string) (uint64, error) {
+	var storageClass storagev1.StorageClass
+	if err := r.client.Get(ctx, types.NamespacedName{Name: name}, &storageClass); err != nil {
+		// suppress error message
+		return 0, err
+	}
+	var replicaCount uint64
+	var err error
+	if storageClass.Provisioner != DriverName {
+		return 0, errors.NewBadRequest(fmt.Sprintf("storageClass provisioner is required to be %s", DriverName))
+	}
+	// maxMountReplicaCount can either be explicitly defined or implicitly deduced from maxShares
+	if maxMountReplicaCount, ok := storageClass.Parameters["maxMountReplicaCount"]; ok {
+		if replicaCount, err = strconv.ParseUint(maxMountReplicaCount, 10, 64); err != nil {
+			klog.Errorf("failed to parse maxMountReplicaCount: (%v)", err)
+			// if the parameter for maxMountReplicaCount is not formatted properly, there is no point in requeueing the request
+			return 0, err
+		}
+	} else if maxShares, ok := storageClass.Parameters["maxShares"]; ok {
+		if replicaCount, err = strconv.ParseUint(maxShares, 10, 64); err != nil {
+			klog.Errorf("failed to parse maxMountReplicaCount: (%v)", err)
+			// if the parameter for maxMountReplicaCount is not formatted properly, there is no point in requeueing the request
+			return 0, err
+		}
+		replicaCount--
+		// if maxShares field does not exist, the storage class is not supported by azVolumeAttachment Controller
+	} else {
+		klog.Errorf("unable to find maxShares in storage class %s's parameters", name)
+		return 0, errors.NewBadRequest("maxShares field is required")
+	}
+	return replicaCount, nil
+}
+
+func (r *reconcileAzVolumeAttachment) ManageReplicaForVolume(ctx context.Context, volume string) error {
+	// fetch the replica azVolumeAttachments for the specified underlying volume
+	replicas, ok := r.replicaMap[volume]
+	if !ok {
+		r.replicaMap[volume] = make(map[string]v1alpha1.AzVolumeAttachment)
+		replicas = r.replicaMap[volume]
+	}
+	storageClassName, ok := r.volMap[volume]
+	if !ok {
+		// this warning message should never be reached, otherwise, there is a sync issue with the maps used for optimization
+		klog.Warning(fmt.Sprintf("storage class not registerd for volume %s.", volume))
+		return nil
+	}
+	if replicaCount, err := r.GetReplicaCount(ctx, storageClassName); err != nil {
```

Comment:
It would be clearer to rename `replicaCount` to `desiredReplicaCount`.

Reply:
Done